### PR TITLE
stdlib: fix signatures for some functions with unrepresentable defaults

### PR DESCRIPTION
**Pull request by JelleZijlstra**
Found with python/mypy#16433

### CODE REVIEW

File: stdlib/_locale.pyi [hunk 1]
```diff
@@ -10,7 +10,7 @@ LC_NUMERIC: int
 LC_ALL: int
 CHAR_MAX: int
 
-def setlocale(category: int, locale: str | Iterable[str | None] | None = None) -> str: ...
+def setlocale(__category: int, __locale: str | Iterable[str | None] | None = None) -> str: ...
```
Comment (AlexWaygood) [inline stdlib/_locale.pyi:13]: This works fine with keyword args on my machine (on Python 3.8 and Python 3.13):

```pycon
>>> import locale
>>> locale.setlocale(0, locale='foo')
'foo'
>>> locale.setlocale(category=0, locale='foo')
'foo'
```

Comment (AlexWaygood) [reply stdlib/_locale.pyi:13]: I see CI is failing for this one as well :)

Comment (JelleZijlstra) [reply stdlib/_locale.pyi:13]: That's `locale`, not `_locale`. `_locale.setlocale` really takes only positional args: https://github.com/python/cpython/blob/cc18b886a51672c59622837a2b8e83bf6be28c58/Modules/_localemodule.c#L91. But of course people should normally use the `locale` version, and stubtest is also complaining here. I'll look into how to fix this.

Comment (AlexWaygood) [reply stdlib/_locale.pyi:13]: Oh, right. Maybe we re-export `_locale.setlocale` in `locale.setlocale` currently, whereas at runtime they're actually two different functions?

Comment (AlexWaygood) [reply stdlib/_locale.pyi:13]: Indeed, we re-export it in the stub: https://github.com/python/typeshed/blob/7f9b3ea6c354273ff6ef78c15f274d6e29becb22/stdlib/locale.pyi#L11

But that doesn't reflect what happens at runtime, where `locale.setlocale()` is a pure-Python wrapper around `_locale.setlocale()`: https://github.com/python/cpython/blob/cc18b886a51672c59622837a2b8e83bf6be28c58/Lib/locale.py#L600-L615

Comment (JelleZijlstra) [reply stdlib/_locale.pyi]: Fixed. Also noticed that the argument type for `_locale.setlocale` was wrong; unlike the higher-level version it doesn't accept `Iterable[str]`.

File: stdlib/locale.pyi [hunk 2]
```diff
@@ -131,6 +130,7 @@ def getdefaultlocale(
     envvars: tuple[_str, ...] = ("LC_ALL", "LC_CTYPE", "LANG", "LANGUAGE")
 ) -> tuple[_str | None, _str | None]: ...
 def getlocale(category: int = ...) -> tuple[_str | None, _str | None]: ...
+def setlocale(__category: int, locale: _str | Iterable[_str | None] | None = None) -> _str: ...
```
Comment (AlexWaygood) [inline stdlib/locale.pyi:133]:
```suggestion
def setlocale(category: int, locale: _str | Iterable[_str | None] | None = None) -> _str: ...
```

**Review by AlexWaygood — APPROVED** [review]
These all look correct.

I pick up 14 more errors running stdlib stubtest locally on Python 3.12 with a Windows machine with your mypy PR branch -- you could do those as well? But I'm also happy to defer those to another PR

<details>
<summary>Additional errors</summary>

```
error: _warnings.warn is inconsistent, stub does not have argument "skip_file_prefixes"
Stub: in file stdlib\_warnings.pyi:7
Overload(def (message: builtins.str, category: Union[Type[builtins.Warning], None] =, stacklevel: builtins.int =, source: Union[Any, None] =), def (message: builtins.Warning, category: Any =, stacklevel: builtins.int =, source: Union[Any, None] =))
Inferred signature: def (message: Union[builtins.str, builtins.Warning], category: Union[Type[builtins.Warning], None, Any] = ..., stacklevel: builtins.int = ..., source: Union[Any, None] = ...)
Runtime:
def (message, category=None, stacklevel=1, source=None, *, skip_file_prefixes=<unrepresentable>)

error: builtins.getattr is inconsistent, stub argument "__name" has a default value but runtime argument does not
Stub: in file stdlib\builtins.pyi:1426
Overload(def (builtins.object, builtins.str) -> Any, def (builtins.object, builtins.str, None) -> Union[Any, None], def (builtins.object, builtins.str, builtins.bool) -> Union[Any, builtins.bool], def (builtins.object, name: builtins.str, builtins.list[Any]) -> Union[Any, builtins.list[Any]], def (builtins.object, name: builtins.str, builtins.dict[Any, Any]) -> Union[Any, builtins.dict[Any, Any]], def [_T] (builtins.object, builtins.str, _T`-1) -> Union[Any, _T`-1])
Inferred signature: def (__o: builtins.object, __name: builtins.str = ..., name: builtins.str = ..., __default: Union[None, builtins.bool, builtins.list[Any], builtins.dict[Any, Any], _T`-1] = ...)
Runtime:
def (object, name, default=<unrepresentable>, /)

error: builtins.getattr is inconsistent, stub argument "name" differs from runtime argument "default"
Stub: in file stdlib\builtins.pyi:1426
Overload(def (builtins.object, builtins.str) -> Any, def (builtins.object, builtins.str, None) -> Union[Any, None], def (builtins.object, builtins.str, builtins.bool) -> Union[Any, builtins.bool], def (builtins.object, name: builtins.str, builtins.list[Any]) -> Union[Any, builtins.list[Any]], def (builtins.object, name: builtins.str, builtins.dict[Any, Any]) -> Union[Any, builtins.dict[Any, Any]], def [_T] (builtins.object, builtins.str, _T`-1) -> Union[Any, _T`-1])
Inferred signature: def (__o: builtins.object, __name: builtins.str = ..., name: builtins.str = ..., __default: Union[None, builtins.bool, builtins.list[Any], builtins.dict[Any, Any], _T`-1] = ...)
Runtime:
def (object, name, default=<unrepresentable>, /)

error: builtins.getattr is inconsistent, stub argument "name" should be positional-only (rename with a leading double underscore, i.e. "__default")
Stub: in file stdlib\builtins.pyi:1426
Overload(def (builtins.object, builtins.str) -> Any, def (builtins.object, builtins.str, None) -> Union[Any, None], def (builtins.object, builtins.str, builtins.bool) -> Union[Any, builtins.bool], def (builtins.object, name: builtins.str, builtins.list[Any]) -> Union[Any, builtins.list[Any]], def (builtins.object, name: builtins.str, builtins.dict[Any, Any]) -> Union[Any, builtins.dict[Any, Any]], def [_T] (builtins.object, builtins.str, _T`-1) -> Union[Any, _T`-1])
Inferred signature: def (__o: builtins.object, __name: builtins.str = ..., name: builtins.str = ..., __default: Union[None, builtins.bool, builtins.list[Any], builtins.dict[Any, Any], _T`-1] = ...)
Runtime:
def (object, name, default=<unrepresentable>, /)

error: builtins.getattr is inconsistent, runtime does not have argument "__default"
Stub: in file stdlib\builtins.pyi:1426
Overload(def (builtins.object, builtins.str) -> Any, def (builtins.object, builtins.str, None) -> Union[Any, None], def (builtins.object, builtins.str, builtins.bool) -> Union[Any, builtins.bool], def (builtins.object, name: builtins.str, builtins.list[Any]) -> Union[Any, builtins.list[Any]], def (builtins.object, name: builtins.str, builtins.dict[Any, Any]) -> Union[Any, builtins.dict[Any, Any]], def [_T] (builtins.object, builtins.str, _T`-1) -> Union[Any, _T`-1])
Inferred signature: def (__o: builtins.object, __name: builtins.str = ..., name: builtins.str = ..., __default: Union[None, builtins.bool, builtins.list[Any], builtins.dict[Any, Any], _T`-1] = ...)
Runtime:
def (object, name, default=<unrepresentable>, /)

error: builtins.iter is inconsistent, stub argument "__iterable" differs from runtime argument "object"
Stub: in file stdlib\builtins.pyi:1453
Overload(def [_SupportsNextT <: _typeshed.SupportsNext[Any]] (_typeshed.SupportsIter[_SupportsNextT`-1]) -> _SupportsNextT`-1, def [_T] (builtins._GetItemIterable[_T`-1]) -> typing.Iterator[_T`-1], def [_T] (def () -> Union[_T`-1, None], None) -> typing.Iterator[_T`-1], def [_T] (def () -> _T`-1, builtins.object) -> typing.Iterator[_T`-1])
Inferred signature: def (__iterable: Union[_typeshed.SupportsIter[_SupportsNextT`-1], builtins._GetItemIterable[_T`-1], def () -> Union[_T`-1, None]], __sentinel: builtins.object = ...)
Runtime:
def (object, sentinel=<unrepresentable>, /)

error: codecs.StreamReader.__getattr__ is inconsistent, runtime argument "getattr" has a default value of type def (Any, Any, Any =) -> Any, which is incompatible with stub argument type def (builtins.str) -> Any
Stub: in file stdlib\codecs.pyi:230
def (codecs.StreamReader, builtins.str, def (builtins.str) -> Any =) -> Any
Runtime: in file <frozen codecs>
def (self, name, getattr=<built-in function getattr>)

error: codecs.StreamWriter.__getattr__ is inconsistent, runtime argument "getattr" has a default value of type def (Any, Any, Any =) -> Any, which is incompatible with stub argument type def (builtins.str) -> Any
Stub: in file stdlib\codecs.pyi:216
def (codecs.StreamWriter, builtins.str, def (builtins.str) -> Any =) -> Any
Runtime: in file <frozen codecs>
def (self, name, getattr=<built-in function getattr>)

error: collections.OrderedDict.pop is inconsistent, stub argument "__key" should be positional or keyword (remove leading double underscore)
Stub: in file stdlib\collections\__init__.pyi:1106
Overload(def (self: builtins.dict[_KT`1, _VT`2], _KT`1) -> _VT`2, def (self: builtins.dict[_KT`1, _VT`2], _KT`1, _VT`2) -> _VT`2, def [_T] (self: builtins.dict[_KT`1, _VT`2], _KT`1, _T`-1) -> Union[_VT`2, _T`-1])
Inferred signature: def (self: builtins.dict[_KT`1, _VT`2], __key: _KT`1, __default: Union[_VT`2, _T`-1] = ...)
Runtime:
def (self, /, key, default=<unrepresentable>)

error: collections.OrderedDict.pop is inconsistent, stub argument "__default" should be positional or keyword (remove leading double underscore)
Stub: in file stdlib\collections\__init__.pyi:1106
Overload(def (self: builtins.dict[_KT`1, _VT`2], _KT`1) -> _VT`2, def (self: builtins.dict[_KT`1, _VT`2], _KT`1, _VT`2) -> _VT`2, def [_T] (self: builtins.dict[_KT`1, _VT`2], _KT`1, _T`-1) -> Union[_VT`2, _T`-1])
Inferred signature: def (self: builtins.dict[_KT`1, _VT`2], __key: _KT`1, __default: Union[_VT`2, _T`-1] = ...)
Runtime:
def (self, /, key, default=<unrepresentable>)

error: contextvars.ContextVar.get is inconsistent, stub argument "default" should be positional-only (rename with a leading double underscore, i.e. "__default")
Stub: in file stdlib\contextvars.pyi:24
Overload(def (self: contextvars.ContextVar[_T`1]) -> _T`1, def (self: contextvars.ContextVar[_T`1], default: _T`1) -> _T`1, def [_D] (self: contextvars.ContextVar[_T`1], default: _D`-1) -> Union[_D`-1, _T`1])
Inferred signature: def (self: contextvars.ContextVar[_T`1], default: Union[_T`1, _D`-1] = ...)
Runtime:
def (self, default=<unrepresentable>, /)

error: sqlite3.Connection.cursor is inconsistent, stub argument "cursorClass" differs from runtime argument "factory"
Stub: in file stdlib\sqlite3\__init__.pyi:351
Overload(def (self: sqlite3.dbapi2.Connection, cursorClass: None =) -> sqlite3.dbapi2.Cursor, def [_CursorT <: sqlite3.dbapi2.Cursor] (self: sqlite3.dbapi2.Connection, cursorClass: def (sqlite3.dbapi2.Connection) -> _CursorT`-1) -> _CursorT`-1)
Inferred signature: def (self: sqlite3.dbapi2.Connection, cursorClass: Union[None, def (sqlite3.dbapi2.Connection) -> _CursorT`-1] = ...)
Runtime:
def (self, /, factory=<unrepresentable>)

error: sqlite3.dbapi2.Connection.cursor is inconsistent, stub argument "cursorClass" differs from runtime argument "factory"
Stub: in file stdlib\sqlite3\dbapi2.pyi:351
Overload(def (self: sqlite3.dbapi2.Connection, cursorClass: None =) -> sqlite3.dbapi2.Cursor, def [_CursorT <: sqlite3.dbapi2.Cursor] (self: sqlite3.dbapi2.Connection, cursorClass: def (sqlite3.dbapi2.Connection) -> _CursorT`-1) -> _CursorT`-1)
Inferred signature: def (self: sqlite3.dbapi2.Connection, cursorClass: Union[None, def (sqlite3.dbapi2.Connection) -> _CursorT`-1] = ...)
Runtime:
def (self, /, factory=<unrepresentable>)

error: warnings.warn is inconsistent, stub does not have argument "skip_file_prefixes"
Stub: in file stdlib\warnings.pyi:7
Overload(def (message: builtins.str, category: Union[Type[builtins.Warning], None] =, stacklevel: builtins.int =, source: Union[Any, None] =), def (message: builtins.Warning, category: Any =, stacklevel: builtins.int =, source: Union[Any, None] =))
Inferred signature: def (message: Union[builtins.str, builtins.Warning], category: Union[Type[builtins.Warning], None, Any] = ..., stacklevel: builtins.int = ..., source: Union[Any, None] = ...)
Runtime:
def (message, category=None, stacklevel=1, source=None, *, skip_file_prefixes=<unrepresentable>)

Found 14 errors (checked 726 modules)
```

</details>

**Comment by JelleZijlstra** [comment]
I was planning to point stubtest at my branch in a separate PR so I could see the fallout across versions and OSes.

Also the `getattr()` error in your output looks like a bug.

**Comment by AlexWaygood** [comment]
> Also the `getattr()` error in your output looks like a bug.

yes

**Pull request merged**